### PR TITLE
Compare DDP static graph (C++ core) with legacy DDP forward and backward delay.

### PR DESCRIPTION
**Pull request by bowangbj**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#61507**

Summary:

Benchmark Python-only DDP vs production C++ based DistributedDataParallel.
- Implemented a pure python DDP: PythonDDP with support of SYNC and ASYNC reduction
- Added compare_ddp to measure the difference in forward and backward step

Kudos on Shen and Yi for the great idea.

Test Plan:

Test on DevGPUS with 2 CUDA devices.

$python compare_ddp.py

Python only DDP has slightly better (-1%) forward performance and slightly slower (2%-20%) backward performance.
This suggested that we need to keep C++ Core since the maximum latency increase can be 20%. See README.md for details.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29685364](https://our.internmc.facebook.com/intern/diff/D29685364)

### CODE REVIEW

**Comment by bowangbj** [comment]
@bowangbj has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29685364).

**Comment by bowangbj** [comment]
@bowangbj has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29685364).

**Comment by bowangbj** [comment]
@bowangbj has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29685364).

**Comment by bowangbj** [comment]
@bowangbj has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29685364).

**Comment by bowangbj** [comment]
@bowangbj has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29685364).

**Comment by bowangbj** [comment]
Hey Shen, Yi 

The PR is ready for review. I've added two Python-only impls (Legacy DistributedDataParallel from FAIR, pure Python DDP which support async and sync reduction) and compared the fwd and bwd delays, pls see PR summary and compare_ddp_sample.md for test output.  

Key findings: 
1) Good news the Python only Async reduction looks 'slightly' slower than production DDP (as expected). 
2) Async version is much faster than sync version. My impl is similar as legacy impl from FAIR. 

Note that keeping the legacy impl from FAIR here just for reference, pls let me know whether we need to exclude it from this PR.  Looking forward to your cmts.

CC-ed Jerry as FYI.

File: benchmarks/distributed/ddp/compare_ddp.py [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+        # Forward
+        start.record()
+        outputs = ddp_model(inputs)
+        end.record()
+        torch.cuda.synchronize()
+        metrics[MODEL_FORWARD].append(start.elapsed_time(end))
+
+        # Backward
+        start.record()
+        loss_fn(outputs, labels).backward()
```
Comment (mrshenli) [inline benchmarks/distributed/ddp/compare_ddp.py:109]: nit: `loss_fn()` should be part of forward.

File: benchmarks/distributed/ddp/compare_ddp.py [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+        # Reduce all grad, this is needed for non-DDP_CPP_CORE since the hook
+        # for all_reduce does not exist yet.
+        if ddp_option != DDPOption.DDP_CPP_CORE:
+            ddp_model.all_reduce_grads()
+        optimizer.step()
+        optimizer.zero_grad()
```
Comment (mrshenli) [inline benchmarks/distributed/ddp/compare_ddp.py:115]: Let's exclude optimizer from the backward pass

File: benchmarks/distributed/ddp/compare_ddp.py [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+        end.record()
+        torch.cuda.synchronize()
+        metrics[MODEL_BACKWARD].append(start.elapsed_time(end))
+
+    if rank == 0:
+        print(f'\nMetrics for GPU {rank} ddp_option {ddp_option}:')
+        for step, elapsed_milliseconds in metrics.items():
+            A = np.array(elapsed_milliseconds)
+            print(' {N} iterations, {event}, mean={mean} ms, median={median} ms, p90={p90} ms, p99={p99} ms'.format(
```
Comment (mrshenli) [inline benchmarks/distributed/ddp/compare_ddp.py:124]: Shall we exclude the 1st 20 iterations to get rid of CUDACachingAllocator warmup overhead?

File: benchmarks/distributed/ddp/compare_ddp_sample.md [hunk 4]
```diff
@@ -0,0 +1,23 @@
+Sample Output for compare_ddp.py
+
+Key takeaway:
+- forward performance is almost the same. DDPCore is the slowest.
+- async PythonDDP is slightly slower than DDPCore. DDP_CPP_CORE > PYTHON_DDP_ASYNC > PYTHON_DDP_SYNC > LEGACY_DISTRIBUTED_DATA_PARALLEL
+
+```
+Metrics for GPU 0 ddp_option DDPOption.DDP_CPP_CORE:
```
Comment (mrshenli) [inline benchmarks/distributed/ddp/compare_ddp_sample.md:8]: Hey @bowangbj, thanks a lot for putting this comprehensive measurements together within a very short period of time! And the result looks very promising! Shall we try varying the bucket size for both C++ DDP and Python DDP and see how that affects the delay? In theory, this shouldn't make too much difference, as it does not affects the number of C++/Python context switches. But it will be helpful to know that for sure. :)

**Comment by mrshenli** [comment]
> Note that keeping the legacy impl from FAIR here just for reference, pls let me know whether we need to exclude it from this PR. Looking forward to your cmts.

Since your Python DDP implementation is much faster than the legacy one, I'd vote for excluding the legacy one from this PR.

File: benchmarks/distributed/ddp/compare_ddp.py (hunk 2 above)
Comment (bowangbj) [reply benchmarks/distributed/ddp/compare_ddp.py:115]: Done

File: benchmarks/distributed/ddp/compare_ddp.py (hunk 1 above)
Comment (bowangbj) [reply benchmarks/distributed/ddp/compare_ddp.py:109]: Done.  maybe in distributed world, backward is more like "reduction" across processes ?

File: benchmarks/distributed/ddp/compare_ddp.py (hunk 3 above)
Comment (bowangbj) [reply benchmarks/distributed/ddp/compare_ddp.py:124]: Great point. Done.

**Review by mrshenli — COMMENTED** [review]
nit: shall we put the new files into their own folder? (say benchmarks/distributed/ddp/compare) Asking because `benchmarks/distributed/ddp` already has a README.md, which is irrelevant to the benchmark added in this PR.

File: benchmarks/distributed/ddp/compare_ddp.py (hunk 1 above)
Comment (bowangbj) [reply benchmarks/distributed/ddp/compare_ddp.py:109]: After a second thought, it might still make sense to have loss_func.backward in the backward step. 

Otherwise, it seems there will be nothing to measure in the backward step for production DDP?  Just found that the delay for DDP backward is < 1ms since it has nothing to do if loss func is not included :)

Comment (mrshenli) [reply benchmarks/distributed/ddp/compare_ddp.py:109]: oh, this is surprising and doesn't look right. We previously measured things like below, and usually backward is 3x-5x slower than the forward, and it will be even slower when there are communications. See Fig. 6 in http://www.vldb.org/pvldb/vol13/p3005-li.pdf

```python
fwd_start.record()
loss = loss_fn(model(inputs), labels)
fwd_stop.record()

bwd_start.record()
loss.backward()
bwd_stop.record()
```

If you consistently see this behavior (backward delay is negligible compared to forward), could you please share a small repro?

Comment (bowangbj) [reply benchmarks/distributed/ddp/compare_ddp.py:109]: Sry I misread your cmt. 

I misplaced 
loss_fn(outputs, labels).backward()   -- in forward step

forward: 
outputs = model(inputs)
loss_fn(outputs, labels).backward()

backward:
nothing    :)

Changed it back to sth like 
forward:
outputs = models(inputs)
loss = loss_fn(labels, outputs)

backward:
loss.backward()

Tried it a few times, moving loss_fn(labels, outputs) would not make a big difference as expected.

File: benchmarks/distributed/ddp/compare_ddp_sample.md (hunk 4 above)
Comment (bowangbj) [reply benchmarks/distributed/ddp/compare_ddp_sample.md:8]: Great suggestion! 

Added the result per different bucket size. Overall Python_DDP is 2-20% slower than DDPCore for different bucket size. As expected, larger bucket size can reduce parallelization thus it introduces small delay as bucket size increases.  Also once the bucket size is quite large, it won't make difference as bucket size continues to increase.

Please see details at sample_output.md

**Comment by bowangbj** [comment]
@bowangbj has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29685364).

**Comment by bowangbj** [comment]
> > Note that keeping the legacy impl from FAIR here just for reference, pls let me know whether we need to exclude it from this PR. Looking forward to your cmts.
> 
> Since your Python DDP implementation is much faster than the legacy one, I'd vote for excluding the legacy one from this PR.

SG, removed legacy impl

**Comment by bowangbj** [comment]
> nit: shall we put the new files into their own folder? (say benchmarks/distributed/ddp/compare) Asking because `benchmarks/distributed/ddp` already has a README.md, which is irrelevant to the benchmark added in this PR.

SG, moved to compare dir

**Comment by bowangbj** [comment]
Thx Shen for the super quick and great review. Resolved all the cmts, PTAL

**Comment by bowangbj** [comment]
@bowangbj has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29685364).

File: benchmarks/distributed/ddp/compare/compare_ddp.py [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+from enum import Enum
+from tabulate import tabulate
+from torch.nn.parallel import DistributedDataParallel as DDP
+
+class DDPOption(Enum):
+    DDP_CPP_CORE = 1
+    PYTHON_DDP_SYNC_REDUCTION = 2
+    PYTHON_DDP_ASYNC_REDUCTION = 3
+
+class LatencyData:
```
Comment (mrshenli) [inline benchmarks/distributed/ddp/compare/compare_ddp.py:40]: Shall we fix this lint error?

```
[B903] Data class should either be immutable or use __slots__ to save memory. Use collections.namedtuple to generate an immutable class, or enumerate the attributes in a __slot__ declaration in the class to leave attributes mutable.
```

File: benchmarks/distributed/ddp/compare/compare_ddp.py [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+                append_delta(row, np.mean(A_baseline), np.mean(A_exp))
+                for px in [50, 90, 95, 99]:
+                    base = np.percentile(A_baseline, px)
+                    exp = np.percentile(A_exp, px)
+                    row.append(exp)
+                    append_delta(row, base, exp)
+                data.append(row)
+
+            # Output buffer_size, step as a table.
+            print (tabulate(data,
```
Comment (mrshenli) [inline benchmarks/distributed/ddp/compare/compare_ddp.py:217]: Shall we fix lint errors in this file as well?

File: benchmarks/distributed/ddp/compare/python_ddp.py [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+    """
+    Python only implementation for DistributedDataParallel module.
+
+    Unlike the production DistributedDataParallel which relies on many C++ core
+    utils to manage gradient distribution and reduction. This class implement
+    all functions in pure Python such as param bucketing, gradient
+    synchronization and reduction. The only C++ dependency is the common utils:
+    ``dist.all_reduce``
+
+    The idea: parallize gradient calculation and reduction, the same algo as
```
Comment (mrshenli) [inline benchmarks/distributed/ddp/compare/python_ddp.py:15]: parallize -> parallelize

**Review by mrshenli — APPROVED** [review]
LGTM! Thanks for conducting comprehensive experiments! These experiments suggested that we should keep the C++ core, as 20% speed regression is too much.

**Comment by mrshenli** [comment]
Please fix all lint errors before landing. Thanks!

File: benchmarks/distributed/ddp/compare/compare_ddp.py (hunk 6 above)
Comment (bowangbj) [reply benchmarks/distributed/ddp/compare/compare_ddp.py:217]: Done

File: benchmarks/distributed/ddp/compare/python_ddp.py (hunk 7 above)
Comment (bowangbj) [reply benchmarks/distributed/ddp/compare/python_ddp.py:15]: Great catch

File: benchmarks/distributed/ddp/compare/compare_ddp.py (hunk 5 above)
Comment (bowangbj) [reply benchmarks/distributed/ddp/compare/compare_ddp.py:40]: SG,  updated with __slots__

**Comment by bowangbj** [comment]
@bowangbj has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29685364).

**Comment by bowangbj** [comment]
@bowangbj has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29685364).

**Comment by bowangbj** [comment]
@bowangbj has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29685364).

**Comment by bowangbj** [comment]
Thanks Shen again for the great review. Fixed all the lint errors, landing ...

File: benchmarks/distributed/ddp/compare/python_ddp.py [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+    def __init__(self, module, process_group, async_reduction=True, buffer_size=2 ** 22):
+        super(PythonDDP, self).__init__()
+
+        self.module = module
+        self.process_group = process_group
+        self.world_size = dist.get_world_size(group=self.process_group)
+        self.async_reduction = async_reduction
+
+        # Holds all_reduce handles, used when async_reduction is True
+        self.async_handles = set()
```
Comment (SciPioneer) [inline benchmarks/distributed/ddp/compare/python_ddp.py:103]: Can we try using list instead of a set here? When the bucket size is large, the insertion and scanning of a hash set can be slower than a flat vector.

Instead, use `[None] * bucket_size` should be equivalent to a vector + reserve the size first.

File: benchmarks/distributed/ddp/compare/python_ddp.py [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+        mapping. Returns tuple of (param_to_buckets, buckets).
+        """
+        print("_build_buckets_for_params called")
+        params_to_buckets = {}
+        buckets = set()
+        cur_bucket = self.Bucket(max_buffer_size)
+        total_param = 0
+        for param in self.module.parameters():
+            total_param += 1
+            assert param.requires_grad, "param.requires_grad must be True"
```
Comment (SciPioneer) [inline benchmarks/distributed/ddp/compare/python_ddp.py:129]: Do we have a lot of parameters/layers? If so, `assert` here can slow down the loop, as it requires a sync between host and device.

File: benchmarks/distributed/ddp/compare/python_ddp.py [hunk 10]
```diff
@@ -0,0 +1,228 @@
...
+        bucket. When the bucket is full, kickoff an async all_reduce if
+        async_reduction is set, and adds the resultant handle to
+        self.async_handles.
+
+        .. warning::
+            Note param.grad isn't set yet. Use the passed grad instead.
+        """
+        # Validate bucket and offset are set.
+        bucket = self.param_to_bucket.get(param)
+        assert bucket is not None, "Failed to find bucket for param"
```
Comment (SciPioneer) [inline benchmarks/distributed/ddp/compare/python_ddp.py:170]: The `assert` in this method is also called in a loop.

File: benchmarks/distributed/ddp/compare/python_ddp.py [hunk 11]
```diff
@@ -0,0 +1,228 @@
...
+            else:
+                new_bucket = self.Bucket(max_buffer_size)
+                assert new_bucket.try_hold_param(param), "param must be holded in a empty bucket"
+                params_to_buckets[param] = new_bucket
+                buckets.add(new_bucket)
+                cur_bucket = new_bucket
+
+        first_param = next(self.module.parameters())
+        for bucket in buckets:
+            bucket.buffer = first_param.new(bucket.total_elements)
```
Comment (SciPioneer) [inline benchmarks/distributed/ddp/compare/python_ddp.py:142]: `torch.Tensor.new` is deprecated, and it is likely to copy the data.

It will more efficient to use `new_empty`. See: https://pytorch.org/docs/stable/generated/torch.Tensor.new_empty.html

File: benchmarks/distributed/ddp/compare/python_ddp.py [hunk 12]
```diff
@@ -0,0 +1,228 @@
...
+        otherwise, kicks off synchrous all_reduce for all buckets.
+
+        Once all all buckets are reduced, copy the reduced grads back to their
+        original parameters. After that, reset all buckets in prep for the next
+        iteration.
+        """
+        if self.async_reduction:
+            for handle in self.async_handles:
+                handle.wait()
+            self.async_handles.clear()
```
Comment (SciPioneer) [inline benchmarks/distributed/ddp/compare/python_ddp.py:209]: If `async_handles` is a list of a fixed length, instead of a set, here we can just use `torch.futures.wait_all(async_handles)`, although I don't think it will make a difference in performance.

**Comment by SciPioneer** [comment]
@bowangbj  Did you get a chance to collect the traces to compare the trainings using DDP core and Python Async DDP?

File: benchmarks/distributed/ddp/compare/python_ddp.py (hunk 9 above)
Comment (bowangbj) [reply benchmarks/distributed/ddp/compare/python_ddp.py:129]: there are 1M params. Good idea -- we can introduce debug level to toggle it.

File: benchmarks/distributed/ddp/compare/python_ddp.py (hunk 11 above)
Comment (bowangbj) [reply benchmarks/distributed/ddp/compare/python_ddp.py:142]: Good idea, will address it in follow up CLs.

File: benchmarks/distributed/ddp/compare/python_ddp.py (hunk 10 above)
Comment (bowangbj) [reply benchmarks/distributed/ddp/compare/python_ddp.py:170]: > @bowangbj Did you get a chance to collect the traces to compare the trainings using DDP core and Python Async DDP?

Thanks Yi for all the great comments. Sorry for the long delay. Finally I landed in PyTorch team after bootcamp :) 
Will follow up and dive into more details as part of my onboarding plan in Aug (thanks to Shen, Yi, Boris and Nikhil for the great proposal). 

Stay tuned!
